File: dsa/20-Dynamic Programming/GeeksForGeeks/01Knapsack.cpp
#include <bits/stdc++.h>
using namespace std;

// exponential recursion code
class Solution {
   public:
    int knapsack(int W, vector<int> &val, vector<int> &wt) {
        return helper(val, wt, W, wt.size());
    }

    int helper(vector<int> &val, vector<int> &wt, int W, int n) {
        if (W == 0 || n == 0) {
            return 0;
        }

        if (wt[n - 1] <= W) {
            int val1 = val[n - 1] + helper(val, wt, W - wt[n - 1], n - 1);
            int val2 = helper(val, wt, W, n - 1);

            return max(val1, val2);
        }

        return helper(val, wt, W, n - 1);
    }
};

// memoization - 2D array
class Solution {
   public:
    int knapsack(int W, vector<int> &val, vector<int> &wt) {
        vector<vector<int>> dp(W + 1, vector<int>(wt.size() + 1, -1));
        return helper(val, wt, W, wt.size(), dp);
    }

    int helper(vector<int> &val, vector<int> &wt, int W, int n,
               vector<vector<int>> &dp) {
        if (W == 0 || n == 0) {
            return 0;
        }

        if (dp[W][n] != -1) {
            return dp[W][n];
        }

        if (wt[n - 1] <= W) {
            int val1 = val[n - 1] + helper(val, wt, W - wt[n - 1], n - 1, dp);
            int val2 = helper(val, wt, W, n - 1, dp);

            return dp[W][n] = max(val1, val2);
        }

        return dp[W][n] = helper(val, wt, W, n - 1, dp);
    }
};

File: dsa/20-Dynamic Programming/GeeksForGeeks/01Knapsack.java
class Solution {
    static int knapsack(int W, int val[], int wt[]) {
        int dp[][] = new int[wt.length + 1][W + 1];
        for (int i = 0; i < dp.length; i++) {
            dp[i][0] = 0;
        }

        for (int j = 0; j < dp[0].length; j++) {
            dp[0][j] = 0;
        }

        for (int i = 1; i < dp.length; i++) {
            for (int j = 1; j < dp[0].length; j++) {
                int v = val[i - 1];
                int w = wt[i - 1];

                if (w <= j) {
                    dp[i][j] = Math.max(v + dp[i - 1][j - w], dp[i - 1][j]);
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }

        return dp[wt.length][W];
    }
}
